File: backend/sdk/data_platform.py
"""Data Platform SDK mixin — Volume 65."""

from typing import Any, Dict, Optional


class DataPlatformMixin:
    """Synchronous Data Platform mixin."""

    def dp_create_dataset(self, name: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name}
        for k in ("description", "workspace", "project", "owner", "team", "classification", "region", "status", "retention_days"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/datasets"), data=payload)

    def dp_list_datasets(self, limit: int = 50, **kwargs: Any) -> dict:
        params: Dict[str, Any] = {"limit": limit}
        for k in ("status", "classification", "owner"):
            if k in kwargs and kwargs[k] is not None:
                params[k] = kwargs[k]
        return self.get(self._build_url("/data-platform/datasets"), params=params)

    def dp_get_dataset(self, dataset_id: str) -> dict:
        return self.get(self._build_url(f"/data-platform/datasets/{dataset_id}"))

    def dp_create_source(self, name: str, connector: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name, "connector": connector}
        for k in ("credentials", "region", "classification", "owner", "config"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/sources"), data=payload)

    def dp_create_schema(self, dataset_id: str, fields: list, version: str = "1.0", **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"version": version, "fields": fields}
        for k in ("classification",):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/schemas"), data=payload, params={"dataset_id": dataset_id})

    def dp_create_pipeline(self, name: str, steps: list, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name, "steps": steps}
        for k in ("description", "dependencies", "schedule", "owner", "region", "priority", "resource_limits", "status"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/pipelines"), data=payload)

    def dp_run_pipeline(self, pipeline_id: str, payload: dict | None = None) -> dict:
        return self.post(self._build_url(f"/data-platform/pipelines/{pipeline_id}/runs"), data=payload or {})

    def dp_create_quality_rule(self, dataset_id: str, name: str, rule_type: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name, "rule_type": rule_type}
        for k in ("params", "version"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/quality/rules"), data=payload, params={"dataset_id": dataset_id})

    def dp_run_quality(self, dataset_id: str, records: list) -> dict:
        return self.post(self._build_url("/data-platform/quality/jobs"), data={"dataset_id": dataset_id, "records": records})

    def dp_create_lineage(self, source: str, target: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"source": source, "target": target}
        for k in ("transformation", "pipeline_id", "column_lineage"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/lineage"), data=payload)

    def dp_get_lineage(self, node: str, direction: str = "upstream", depth: int = 3) -> dict:
        return self.get(self._build_url(f"/data-platform/lineage/{node}/{direction}"), params={"depth": depth})

    def dp_search_catalog(self, q: str | None = None, semantic: bool = False, offline: bool = False, limit: int = 20) -> dict:
        params: Dict[str, Any] = {"limit": limit, "semantic": semantic, "offline": offline}
        if q:
            params["q"] = q
        return self.get(self._build_url("/data-platform/catalog/search"), params=params)

    def dp_create_stream(self, topic: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"topic": topic}
        for k in ("partition", "consumer_group", "region"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return self.post(self._build_url("/data-platform/streams"), data=payload)

    def dp_ingest_stream(self, topic: str, payload: dict) -> dict:
        return self.post(self._build_url(f"/data-platform/streams/{topic}/ingest"), data=payload)


class AsyncDataPlatformMixin:
    """Async Data Platform mixin."""

    async def dp_create_dataset(self, name: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name}
        for k in ("description", "workspace", "project", "owner", "team", "classification", "region", "status", "retention_days"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/datasets"), data=payload)

    async def dp_list_datasets(self, limit: int = 50, **kwargs: Any) -> dict:
        params: Dict[str, Any] = {"limit": limit}
        for k in ("status", "classification", "owner"):
            if k in kwargs and kwargs[k] is not None:
                params[k] = kwargs[k]
        return await self.get(self._build_url("/data-platform/datasets"), params=params)

    async def dp_get_dataset(self, dataset_id: str) -> dict:
        return await self.get(self._build_url(f"/data-platform/datasets/{dataset_id}"))

    async def dp_create_source(self, name: str, connector: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name, "connector": connector}
        for k in ("credentials", "region", "classification", "owner", "config"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/sources"), data=payload)

    async def dp_create_schema(self, dataset_id: str, fields: list, version: str = "1.0", **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"version": version, "fields": fields}
        for k in ("classification",):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/schemas"), data=payload, params={"dataset_id": dataset_id})

    async def dp_create_pipeline(self, name: str, steps: list, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name, "steps": steps}
        for k in ("description", "dependencies", "schedule", "owner", "region", "priority", "resource_limits", "status"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/pipelines"), data=payload)

    async def dp_run_pipeline(self, pipeline_id: str, payload: dict | None = None) -> dict:
        return await self.post(self._build_url(f"/data-platform/pipelines/{pipeline_id}/runs"), data=payload or {})

    async def dp_create_quality_rule(self, dataset_id: str, name: str, rule_type: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"name": name, "rule_type": rule_type}
        for k in ("params", "version"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/quality/rules"), data=payload, params={"dataset_id": dataset_id})

    async def dp_run_quality(self, dataset_id: str, records: list) -> dict:
        return await self.post(self._build_url("/data-platform/quality/jobs"), data={"dataset_id": dataset_id, "records": records})

    async def dp_create_lineage(self, source: str, target: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"source": source, "target": target}
        for k in ("transformation", "pipeline_id", "column_lineage"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/lineage"), data=payload)

    async def dp_get_lineage(self, node: str, direction: str = "upstream", depth: int = 3) -> dict:
        return await self.get(self._build_url(f"/data-platform/lineage/{node}/{direction}"), params={"depth": depth})

    async def dp_search_catalog(self, q: str | None = None, semantic: bool = False, offline: bool = False, limit: int = 20) -> dict:
        params: Dict[str, Any] = {"limit": limit, "semantic": semantic, "offline": offline}
        if q:
            params["q"] = q
        return await self.get(self._build_url("/data-platform/catalog/search"), params=params)

    async def dp_create_stream(self, topic: str, **kwargs: Any) -> dict:
        payload: Dict[str, Any] = {"topic": topic}
        for k in ("partition", "consumer_group", "region"):
            if k in kwargs and kwargs[k] is not None:
                payload[k] = kwargs[k]
        return await self.post(self._build_url("/data-platform/streams"), data=payload)

    async def dp_ingest_stream(self, topic: str, payload: dict) -> dict:
        return await self.post(self._build_url(f"/data-platform/streams/{topic}/ingest"), data=payload)

File: frontend/src/types/data-platform.ts
"use client";

/**
 * Data Platform (V65) domain types. Shapes mirror the exact backend
 * serializers in backend/app/api/data_platform.py and
 * backend/app/data_platform/ — never invent fields.
 *
 * Notes:
 * - Most list endpoints return `{items}` with NO `total`; counts derived
 *   client-side are labeled "listed", never presented as totals.
 * - Source credentials are write-only: the backend stores a hash ref and
 *   never returns credential material, so no credential field exists here.
 * - Lakehouse `row_count` values are backend-reported (the raw tier count
 *   is a backend placeholder) and must be labeled as reported, never
 *   treated as measured truth.
 */

export interface PaginatedItems<T> {
  items: T[];
  total?: number;
}

/** POST /data-platform/datasets response + GET /datasets list row. */
export interface DatasetListItem {
  id: string;
  name: string;
  status: string;
  classification: string;
  owner?: string | null;
  region?: string | null;
}

/** GET /data-platform/datasets/{id} response. */
export interface DatasetDetail extends DatasetListItem {
  description?: string | null;
  storage_location?: string | null;
}

/** POST /data-platform/datasets/{id}/versions response. */
export interface DatasetVersion {
  id: string;
  version: string;
  schema_version: string;
}

/** POST /data-platform/sources response. */
export interface DataSourceCreated {
  id: string;
  name: string;
  connector: string;
  status: string;
}

/** GET /data-platform/sources row (no status, no detail endpoint). */
export interface DataSourceListItem {
  id: string;
  name: string;
  connector: string;
  region?: string | null;
}

/** Schema field shape per the DataSchema model: {name, type, nullable, classification}. */
export interface SchemaField {
  name: string;
  type: string;
  nullable?: boolean;
  classification?: string;
}

/** POST /data-platform/schemas response. */
export interface DataSchema {
  id: string;
  version: string;
  fields: SchemaField[];
}

/** GET /data-platform/schemas row. */
export interface DataSchemaListItem {
  id: string;
  dataset_id: string;
  version: string;
  is_published: boolean;
}

/** POST /data-platform/pipelines response. */
export interface DataPipelineCreated {
  id: string;
  name: string;
  status: string;
  dag_hash?: string | null;
}

/** GET /data-platform/pipelines row. */
export interface DataPipelineListItem {
  id: string;
  name: string;
  status: string;
  priority?: string | null;
}

/** POST /data-platform/pipelines/{id}/runs response. */
export interface PipelineRunStarted {
  run_id: string;
  status: string;
  idempotency_key?: string | null;
}

/** POST /data-platform/pipelines/runs/{run_id}/complete response. */
export interface PipelineRunCompleted {
  run_id: string;
  status: string;
  duration_ms?: number | null;
}

/** POST /data-platform/pipelines/{id}/backfill response. */
export interface PipelineBackfill {
  run_id: string;
  status: string;
  scope?: Record<string, unknown>;
  time_range?: Record<string, unknown>;
}

/** GET /data-platform/data-jobs row (run history source). */
export interface DataJobRow {
  run_id: string;
  status: string;
  records?: number | null;
}

/** POST /data-platform/quality/rules response. */
export interface QualityRuleCreated {
  id: string;
  name: string;
  version: string;
}

/** Quality job per-rule counters (POST /quality/jobs + GET /quality/results). */
export interface QualityResultRow {
  rule_id: string;
  passed: number;
  failed: number;
  timestamp?: string;
}

/** POST /data-platform/quality/profile response. */
export interface QualityProfile {
  row_count: number;
  null_rate: Record<string, number>;
  distinct_count: Record<string, number>;
  min_max: Record<string, unknown>;
  distribution?: Record<string, unknown>;
}

/** POST /data-platform/lineage response. */
export interface LineageEdgeCreated {
  id: string;
  source: string;
  target: string;
}

export interface LineageEdge {
  source: string;
  target: string;
  transformation?: string | null;
}

/** GET /data-platform/lineage/{node}/graph response. */
export interface LineageGraph {
  node: string;
  upstream: LineageEdge[];
  downstream: LineageEdge[];
  provenance: unknown[];
}

/** POST /data-platform/streams response. */
export interface DataStreamCreated {
  id: string;
  topic: string;
  partition: number;
}

/** POST /data-platform/streams/{topic}/ingest response. */
export interface StreamEventReceipt {
  event_id: string;
  topic: string;
  partition: number;
  idempotency_key?: string | null;
}

/** GET /data-platform/streams/{topic}/lag response (computed server-side). */
export interface StreamLag {
  topic: string;
  consumer: string;
  lag: number;
}

/** Stream event returned by POST /streams/{topic}/consume. */
export interface StreamEvent {
  event_id?: string;
  topic?: string;
  partition?: number;
  offset?: number;
  payload?: Record<string, unknown>;
}

/** GET /data-platform/lakehouse/{id}/stats response. */
export interface LakehouseTierStat {
  exists: boolean;
  row_count: number;
}

export interface LakehouseStats {
  stats: Record<string, LakehouseTierStat>;
  optimizations: Array<Record<string, unknown>>;
}

/** POST /data-platform/lakehouse/{id}/tier response. */
export interface LakehouseTierWrite {
  tier: string;
  records: number;
  format: string;
  dataset_id: string;
}

export const DATA_CLASSIFICATIONS = ["PUBLIC", "INTERNAL", "CONFIDENTIAL", "RESTRICTED"] as const;
export const DATA_CONNECTORS = [
  "postgresql",
  "object_storage",
  "api",
  "git",
  "csv",
  "json",
  "parquet",
  "event_stream",
] as const;
export const LAKEHOUSE_TIERS = ["raw", "validated", "curated", "serving"] as const;
export const QUALITY_RULE_TYPES = ["required", "range", "regex", "uniqueness", "referential"] as const;
export const SCHEMA_COMPATIBILITY = ["backward", "forward", "full"] as const;
